exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/statements/UpsertStatement.kt

```diff
@@ -61,3 +64,79 @@ open class UpsertStatement<Key : Any>(
+ */
+sealed interface UpsertBuilder {
+    /**
+     * Calls the specified function [body] with an [UpsertBuilder] as its receiver and an [UpdateStatement]
+     * as its argument, allowing values to be assigned to the UPDATE clause of an upsert statement.
+     *
+     * To specify manually that the insert value should be used when updating a column, for example within an expression
+     * or function, invoke `insertValue()` with the desired column as the function argument.
+     *
+     * @sample org.jetbrains.exposed.sql.tests.shared.dml.UpsertTests.testUpsertWithManualUpdateUsingInsertValues
+     */
+    fun onUpdate(body: UpsertBuilder.(UpdateStatement) -> Unit) {
+        val arguments = UpdateStatement((this as InsertStatement<*>).table, null).apply {
+            body.invoke(this@UpsertBuilder, this)
+        }.firstDataSet
+        when (this) {
+            is UpsertStatement<*> -> updateValues.putAll(arguments)
+            is BatchUpsertStatement -> updateValues.putAll(arguments)
+        }
+    }
+
+    /**
+     * Specifies that this column should be updated using the same values that would be inserted if there was
+     * no violation of a unique constraint in an upsert statement.
+     *
+     * @sample org.jetbrains.exposed.sql.tests.shared.dml.UpsertTests.testUpsertWithManualUpdateUsingInsertValues
+     */
+    fun <T> insertValue(column: Column<T>): ExpressionWithColumnType<T> = InsertValue(column, column.columnType)
+
+    private class InsertValue<T>(
+        val column: Column<T>,
+        override val columnType: IColumnType<T & Any>
+    ) : ExpressionWithColumnType<T>() {
+        override fun toQueryBuilder(queryBuilder: QueryBuilder) {
+            val transaction = TransactionManager.current()
+            val functionProvider = getFunctionProvider(transaction.db.dialect)
+            functionProvider.insertValue(transaction.identity(column), queryBuilder)
+        }
+    }
+
+    companion object {
+        /** Returns the [FunctionProvider] for valid upsert statement syntax. */
+        fun getFunctionProvider(dialect: DatabaseDialect): FunctionProvider = when (dialect) {
+            is H2Dialect -> when (dialect.h2Mode) {
+                H2Dialect.H2CompatibilityMode.MariaDB, H2Dialect.H2CompatibilityMode.MySQL -> MysqlFunctionProvider.INSTANCE
+                else -> H2FunctionProvider
+            }
+            else -> dialect.functionProvider
+        }
+    }
+}
+
+/** Returns the columns to be used in the conflict condition of an upsert statement. */
+internal fun UpsertBuilder.getKeyColumns(vararg keys: Column<*>): List<Column<*>> {
+    this as InsertStatement<*>
+    return keys.toList().ifEmpty {
+        table.primaryKey?.columns?.toList() ?: table.indices.firstOrNull { it.unique }?.columns
+    } ?: emptyList()
+}
+
+/** Returns the expressions to be used in the update clause of an upsert statement, along with their insert column reference. */
+internal fun UpsertBuilder.getUpdateExpressions(
```

Comment:
These 2 functions are defined as extension functions of the interface to force them to be internal only, since method's can't be.